File: Java/src/com/dbc/algorithm/Tree/SymmetricTree.java
package com.dbc.algorithm.Tree;

import com.dbc.algorithm.ReferOffer.DuiChengDeErChaShuLcof;

public class SymmetricTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return dfs(root.left, root.right);
    }

    public boolean dfs(TreeNode node1, TreeNode node2) {
        if ((node1 != null && node2 == null) || (node2 != null && node1 == null)) {
            return false;
        } else if (node1 == null) {
            return true;
        }
        return (node1.val == node2.val) && dfs(node1.left, node2.right) && dfs(node1.right, node2.left);
    }
}

File: Python/tree/symmetric_tree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def isSymmetric(root: TreeNode) -> bool:
    def isSym(node1: TreeNode, node2: TreeNode) -> bool:
        if (not node1 and node2) or (not node2 and node1):
            return False
        elif not node1 and not node2:
            return True
        
        
        return (node1.val == node2.val) and isSym(node1.left, node2.right) and isSym(node1.right, node2.left)
    if not root: return True
    return isSym(root.left, root.right)
